File: BinaryTree/Leetcode106.java
import java.util.HashMap;
import java.util.Map;

public class Leetcode106 {
    Map<Integer,Integer> map = new HashMap<>();
    public TreeNode buildTree(int[] postorder, int[] inorder){
        for(int i = 0;i < inorder.length;i++){
            map.put(inorder[i],i);
        }
        return helper(0,postorder.length-1,postorder,0,inorder.length-1,inorder);
    }
    private TreeNode helper(int pStart, int pEnd, int[] postorder,int iStart,int iEnd, int[] inorder){
        if(pStart > pEnd || iStart > iEnd)return null;
        int index = map.get(postorder[pEnd]);
        int leftNum = index-iStart;
        TreeNode root = new TreeNode(postorder[pEnd]);
        root.left = helper(pStart,pStart+leftNum-1,postorder,iStart,index,inorder);
        root.right = helper(pStart+leftNum,pEnd-1,postorder,index+1,iEnd,inorder);
        return root;
    }
}

File: BinarySearchTree/Python/Leetcode106.py
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if len(postorder) == 0:
            return None
        node = TreeNode(postorder[-1])
        index = inorder.index(postorder[-1])
        postorderLeft = []
        if index > 0: 
            postorderLeft = postorder[0:index]
            node.left = self.buildTree(inorder[:index], postorderLeft)
        if index < len(inorder) - 1: 
            postorderRight = postorder[len(postorderLeft):-1]
            node.right = self.buildTree(inorder[index + 1:], postorderRight)
        return node
